File: src/2025-03-04-03-28-40--deepseek-chat-0304/java/exercises/practice/simple-linked-list/src/main/java/SimpleLinkedList.java
class SimpleLinkedList<T> {
    private static class Node<T> {
        private T data;
        private Node<T> next;
        
        Node(T data) {
            this.data = data;
            this.next = null;
        }
    }
    
    private Node<T> head;
    private int size;
    
    SimpleLinkedList() {
        this.head = null;
        this.size = 0;
    }

    SimpleLinkedList(T[] values) {
        this();
        for (T value : values) {
            push(value);
        }
    }

    void push(T value) {
        Node<T> newNode = new Node<>(value);
        newNode.next = head;
        head = newNode;
        size++;
    }

    T pop() {
        if (head == null) {
            throw new IllegalStateException("Cannot pop from an empty list");
        }
        T value = head.data;
        head = head.next;
        size--;
        return value;
    }

    void reverse() {
        Node<T> prev = null;
        Node<T> current = head;
        Node<T> next = null;
        
        while (current != null) {
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        head = prev;
    }

    @SuppressWarnings("unchecked")
    T[] asArray(Class<T> clazz) {
        if (size == 0) {
            return (T[]) java.lang.reflect.Array.newInstance(clazz, 0);
        }
        T[] array = (T[]) java.lang.reflect.Array.newInstance(clazz, size);
        Node<T> current = head;
        int index = 0;
        while (current != null) {
            array[index++] = current.data;
            current = current.next;
        }
        return array;
    }

    int size() {
        return size;
    }
}

File: src/python-2025-02-27-deepclaude37-rust/python/exercises/practice/simple-linked-list/simple_linked_list.py
class EmptyListException(Exception):
    def __init__(self, message):
        super().__init__(message)


class Node:
    def __init__(self, value):
        self._value = value
        self._next = None

    def value(self):
        return self._value

    def next(self):
        return self._next


class LinkedList:
    def __init__(self, values=None):
        self.head_node = None
        self._length = 0
        if values is not None:
            for value in values:
                self.push(value)

    def __iter__(self):
        current = self.head_node
        while current is not None:
            yield current.value()
            current = current.next()

    def __len__(self):
        return self._length

    def head(self):
        if self.head_node is None:
            raise EmptyListException("The list is empty.")
        return self.head_node

    def push(self, value):
        new_node = Node(value)
        new_node._next = self.head_node
        self.head_node = new_node
        self._length += 1

    def pop(self):
        if self.head_node is None:
            raise EmptyListException("The list is empty.")
        value = self.head_node.value()
        self.head_node = self.head_node.next()
        self._length -= 1
        return value

    def reversed(self):
        reversed_list = LinkedList()
        current = self.head_node
        while current is not None:
            reversed_list.push(current.value())
            current = current.next()
        return reversed_list
